### ISSUE TITLE
DTO with no properties causes null exception

### ISSUE THREAD

**Issue by jenseralmeida** [JonPSmith/EfCore.GenericServices#30]
Hi, 

I have this code in https://gist.github.com/jenseralmeida/a51b55b2eb5d1e42a5ebd3e0ef2bd269 which is failing without a proper message explaining what I did wrong. The exception should clearly state that the DTO does not have any writable property, instead of a NullReferenceException later on the code where an instance value was expected. 

I am pasting the failing method just below.

![image](https://user-images.githubusercontent.com/660075/61165261-09bdd000-a4d3-11e9-8ad2-e6bbbdc22f2d.png)

The first line of the DecodedDto.MatchCtorsAndStaticMethodsToProperties method is the one that could be changed to check and throws if there is no writable property.

![image](https://user-images.githubusercontent.com/660075/61165276-278b3500-a4d3-11e9-8674-41a271939198.png)

I have put the full code for the unit-test copied from the gist above

``` C#
using GenericServices;
using GenericServices.Setup;
using Microsoft.EntityFrameworkCore;
using Xunit;

namespace readonly_property
{
    public class UnitTest1
    {
        [Fact]
        public void Dto_with_no_writable_property_causes_null_exception()
        {
            using (var context = new SomeContext(InMemoryOptions.Create<SomeContext>()))
                context.SetupSingleDtoAndEntities<ImutableDto>();
        }

        private static class InMemoryOptions
        {
            public static DbContextOptions<T> Create<T>(string databaseName = "InMemory") where T: DbContext
                => new DbContextOptionsBuilder<T>()
                    .UseInMemoryDatabase(databaseName: databaseName)
                    .Options;
        }
    }

    public class SomeContext : DbContext
    {
        public SomeContext(DbContextOptions<SomeContext> options): base(options) { }
        public DbSet<ImutableEntity> Entities { get; set; }
        protected override void OnModelCreating(ModelBuilder builder)
        {
            builder.Entity<ImutableEntity>().Property(e => e.Id).ValueGeneratedNever();
            builder.Entity<ImutableEntity>().Property(e => e.Name);
        }
    }

    public class ImutableEntity
    {
        public ImutableEntity(int id, string name)
        {
            Id = id;
            Name = name;
        }
        public int Id { get; private set; }
        public string Name { get; private set; }
    }

    public class ImutableDto: ILinkToEntity<ImutableEntity>
    {
    }
}
```

**Comment by JonPSmith** [comment]
HI @jenseralmeida,

OK, having a DTO with no properties will cause an error as here must be properties there to match with something. To answer your question on immutable support - yes, it will match the DTO names/types to ctors or public methods in the class - see [DDD entity patterns](https://github.com/JonPSmith/EfCore.GenericServices/wiki/DDD-entity-patterns) docs for how it matches.

PS. You might like to look at the [Unit testing your code](https://github.com/JonPSmith/EfCore.GenericServices/wiki/Unit-testing-your-code) page in the docs - easier way to test things without using DI.

**Comment by jenseralmeida** [comment]
I have changed the title of the issue to "DTO with non-writable properties causes null exception" and I did update the test removing the DI. Thanks for pointing me that.

``` C#
using GenericServices;
using GenericServices.Setup;
using Microsoft.EntityFrameworkCore;
using Xunit;

namespace readonly_property
{
    public class UnitTest1
    {
        [Fact]
        public void Dto_with_no_writable_property_causes_null_exception()
        {
            using (var context = new SomeContext(InMemoryOptions.Create<SomeContext>()))
                context.SetupSingleDtoAndEntities<ImutableDto>();
        }

        private static class InMemoryOptions
        {
            public static DbContextOptions<T> Create<T>(string databaseName = "InMemory") where T: DbContext
                => new DbContextOptionsBuilder<T>()
                    .UseInMemoryDatabase(databaseName: databaseName)
                    .Options;
        }
    }

    public class SomeContext : DbContext
    {
        public SomeContext(DbContextOptions<SomeContext> options): base(options) { }
        public DbSet<ImutableEntity> Entities { get; set; }
        protected override void OnModelCreating(ModelBuilder builder)
        {
            builder.Entity<ImutableEntity>().Property(e => e.Id).ValueGeneratedNever();
            builder.Entity<ImutableEntity>().Property(e => e.Name);
        }
    }

    public class ImutableEntity
    {
        public ImutableEntity(int id, string name)
        {
            Id = id;
            Name = name;
        }
        public int Id { get; private set; }
        public string Name { get; private set; }
    }

    public class ImutableDto: ILinkToEntity<ImutableEntity>
    {
    }
}

```

**Comment by jenseralmeida** [comment]
Hi @JonPSmith ,

I am bringing the following points from the comment https://github.com/JonPSmith/EfCore.GenericServices/issues/31#issuecomment-514642221 to this context, as the questions were related to this issue. Before discussing your points, let me give to you an overall overview of what I have done.

This specific issue is about me trying to use an invalid DTO mapped to an immutable class, and not getting a meaningful message from the framework. Instead, the code blows with a null reference exception. The opened PR brings the same behaviour from a valid DTO, but with a non-matching ctor/method at the immutable entity exception, to an invalid DTO mapped to an immutable entity.

I will be adding specific comments to the relevant parts of the PR and answering overall questions just below.

> I don't think your PR wouldn't cover all the usages of the ParametersMatch class. 

I would love to learn more and making myself understand better the code to be able to do better contributions. If you could point me what did I miss, please, let me know so I can learn and improve.

> Have you run all the unit tests on your version? Do they all pass with your changes?

Yes, all of the tests are passing.

> Can I understand what you are wanting to do.

I have given a long answer above, but the short answer is, I do want to have a meaningful message when I or my colleagues make the same mistake in the future.

**Comment by JonPSmith** [comment]
This is fixed in version 2.0.3

Note that the issue was having NO properties in the DTO, not that they were read only.
